Type-1 font subsetting

jkseppan opened the pull request:
## PR Summary

Type-1 subsetting

This reduces pdf file sizes when usetex is active, at the cost of
some complexity in the code. We implement a charstring bytecode
interpreter to keep track of subroutine calls in font programs.

Recommend merging to main to give people time to test this, not to
a 3.10 point release.

Give dviread.DviFont a fake filename attribute and a get_fontmap
method for character tracking.

Add type hints to the code this touches.

Closes #127.

## PR Checklist

<!-- Please mark any checkboxes that do not apply to this PR as [N/A]. -->

- [ ] Has pytest style unit tests (and `pytest` passes).
- [ ] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (run `flake8` on changed files to check).
- [ ] New features are documented, with examples if plot related.
- [ ] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [ ] Conforms to Matplotlib style conventions (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).
- [ ] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [ ] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).

<!--
Thank you so much for your PR!  To help us review your contribution, please
consider the following points:

- A development guide is available at https://matplotlib.org/devdocs/devel/index.html.

- Help with git and github is available at
  https://matplotlib.org/devel/gitwash/development_workflow.html.

- Do not create the PR out of master, but out of a separate branch.

- The PR title should summarize the changes, for example "Raise ValueError on
  non-numeric input to set_xlim".  Avoid non-descriptive titles such as
  "Addresses issue #8576".

- The summary should provide at least 1-2 sentences describing the pull request
  in detail (Why is this change required?  What problem does it solve?) and
  link to any relevant issues.

- If you are contributing fixes to docstrings, please pay attention to
  http://matplotlib.org/devel/documenting_mpl.html#formatting.  In particular,
  note the difference between using single backquotes, double backquotes, and
  asterisks in the markup.

We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in.  Please let us know if the reviews are unclear or
the recommended next step seems overly demanding, if you would like help in
addressing a reviewer's comments, or if you have been waiting too long to hear
back on your PR.
-->

anntzer commented:
Is this ready for review, now that #20715 has been merged?

jkseppan commented:
This is failing on Ubuntu 22.04 and Windows but passing on 24.04 and Mac. Here's one failing image (test_usetex_pdf.png, so converted from pdf to png on the test system). 

![test_usetex_pdf](https://github.com/user-attachments/assets/b1deb6e0-939a-4996-8c86-9ac5470e67ef)

This looks like the font is entirely broken. The expected image similarly converted looks like this:

![test_usetex-expected_pdf](https://github.com/user-attachments/assets/52311643-aac3-40fd-8c55-62c8a783150a)

Strangely enough, the generated pdf file looks fine on my Mac.

jkseppan commented:
I can repeat the error running on an Ubuntu 22.04 docker image:

```
GPL Ghostscript 9.55.0 (2021-09-27)
Copyright (C) 2021 Artifex Software, Inc.  All rights reserved.
This software is supplied under the GNU AGPLv3 and comes with NO WARRANTY:
see the file COPYING for details.
Processing pages 1 through 1.
Page 1
   **** Error: can't process embedded font stream,
        attempting to load the font using its name.
               Output may be incorrect.
Querying operating system for font files...
Substituting font Helvetica for MPLAAD+CMR17.
Loading NimbusSans-Regular font from /usr/share/ghostscript/9.55.0/Resource/Font/NimbusSans-Regular... 4467044 2957798 6795968 5451925 4 done.
   **** Error: can't process embedded font stream,
        attempting to load the font using its name.
               Output may be incorrect.
Substituting font Helvetica for MPLAAC+CMR12.
   **** Error: can't process embedded font stream,
        attempting to load the font using its name.
               Output may be incorrect.
Substituting font Helvetica for MPLAAA+CMEX10.
   **** Error: can't process embedded font stream,
        attempting to load the font using its name.
               Output may be incorrect.
Substituting font Helvetica-Oblique for MPLAAB+CMMI12.
Loading NimbusSans-Italic font from /usr/share/ghostscript/9.55.0/Resource/Font/NimbusSans-Italic... 4534308 3175939 7138400 5761600 4 done.
```

The subsetting is wrong in some way that breaks Ghostscript 9.55 but not the viewer in macOS or the newer Ghostscript in Ubuntu 24.04. (Ghostscript 9.56 has a completely rewritten PDF interpreter.)

jkseppan commented:
I hope I found the culprit... I was writing an extra delimiter between the Subrs and the Charstrings when one was already there.

anntzer commented on lib/matplotlib/_type1font.py:
```diff
@@ -768,13 +861,262 @@ def transform(self, effects):
+        while todo - done:
+            glyph = next(iter(todo - done))
+            called_glyphs, called_subrs, _, _ = self._simulate(glyph, [], [])
+            todo.update(called_glyphs)
```
`todo.update(called_glyphs - done)` here and then the loop can just be `while todo: glyph = todo.pop(); ...`?

anntzer commented on lib/matplotlib/_type1font.py:
```diff
@@ -768,13 +861,262 @@ def transform(self, effects):
+
+        Returns
+        -------
+            glyphs : set[str]
```
section should not be indented.

anntzer commented on lib/matplotlib/_type1font.py:
```diff
@@ -35,10 +37,64 @@
 from matplotlib.cbook import _format_approx
 from . import _api
+if T.TYPE_CHECKING:
+    from collections.abc import Iterable
```
We tend not to have inline type hints (and I personally very much don't like them), but there are a few exceptions (e.g. _mathtext.py) so I guess it's up to you whether you want to leave them in.

jkseppan replied:
Right – I saw that there are some files with type hints, and figured that the project might be in the process of adding them. I've found type hints pretty useful at work, but of course we should have a consistent style in the project. Has this been discussed in the past on the dev list or somewhere?

anntzer replied:
I'm probably among the most vocal opponents, but I'll defer to @QuLogic or @ksunden on this aspect.

jkseppan replied:
and @tacaswell responded

> It is worth re-opening the discussion of in-line type hints
the world seems to have stabilized and in other projects I have had type hints catch actual bugs before I ran the code (but I have also had to spend 15 minutes trying to sort out how to placate the type checker for code that clearly works a couple of times)


The type hints were useful for me while working on this, since VS Code pointed out some obvious mistakes in real time. But I agree that it is not ideal to leave the file only partially annotated, since it does not pass a full type check in its current state.

I could remove the extra type hints from this PR and make a separate PR to add hints to the whole file?

anntzer commented on lib/matplotlib/_type1font.py:
```diff
@@ -768,13 +861,262 @@ def transform(self, effects):
+
+        Parameters
+        ----------
+        characters : sequence of bytes
```
of bytes, of ord(bytes) (ASCII codepoints -- note that you typed this as Iterable[int]), or of indices into the encoding array? (e.g. how to interpret this for non-standard encodings?)

anntzer reviewed (approved):
A few minor points to be considered, but overall this looks great.

anntzer commented on lib/matplotlib/_type1font.py:
```diff
@@ -768,13 +861,262 @@ def transform(self, effects):
+            postscript_stack: list[float],
+            opcode: int | str,
+        ) -> tuple[set, set, list[float], list[float]]:
+        """Run one step in the charstring interpreter."""
```
I wonder if this may be clearer if you mutate buildchar_stack and postscript_stack in-place?  this way you would write
```python
glyphs = set(); subrs = set()
if opcode in {...}:
    buildchar_stack[:] = []
elif opcode == "seac":
    codes = ...; glyphs.update(...)
    buildchar_stack[:] = []
elif opcode == "div":
    num2 = buildchar_stack.pop()
    num1 = buildchar_stack.pop()
    buildchar_stack.append(num1 / num2)
...
return glyphs, subrs
```
which feels perhaps more in the spirit of a postscript interpreter?

jkseppan replied:
Could be! I'll think about this a little.

jkseppan replied:
I moved this into a separate class with the stacks and glyph/subr sets as members. I agree that it looks clearer that way.

anntzer commented on lib/matplotlib/_type1font.py:
```diff
@@ -35,10 +37,64 @@
+
+        Useful for adding a tag into subsetted fonts while keeping the code
+        reproducible. The function always returns the same value for the
+        same set on the same exact Python version and is free of collisions.
```
"always returns the same value..." -> if the sequence of calls is the same.

jkseppan replied:
Yeah, and if we want reproducible output we shouldn't really make this a global variable. I suppose the tagger should be a part of the PDF file object so that it can ensure that tags within the same file are consistent.

jkseppan commented on lib/matplotlib/_type1font.py:
```diff
@@ -35,10 +37,64 @@
+        self._next = ''.join(chars)
+
+
+_make_tag = _Tagger()
```
- itertools.product: Thanks! I was convinced there is something suitable in itertools but didn't think of that.
- `MPL{num}+`: the spec (PDF reference, 5.5.3 Font Subsets; I'm looking at PDF 1.7 but I suppose the section numbering should be stable across versions) says the tag has to be exactly six uppercase letters.

anntzer replied:
Perhaps add a ref to the relevant section then?

jkseppan replied:
There was similar enough code in the PDF backend, so using that instead.

QuLogic commented on lib/matplotlib/_type1font.py:
```diff
@@ -768,13 +861,262 @@ def transform(self, effects):
+            return glyphs, set(), [], postscript_stack
+        elif opcode == 'div':
+            num1, num2 = buildchar_stack[-2:]
+            return (
+                set(),
+                set(),
+                buildchar_stack[-2:] + [num1/num2], postscript_stack
+            )
```
This fits in a line?
```suggestion
            return set(), set(), buildchar_stack[-2:] + [num1/num2], postscript_stack
```

QuLogic commented on lib/matplotlib/_type1font.py:
```diff
@@ -768,13 +861,262 @@ def transform(self, effects):
+
+    def _subset_encoding(self, encoding: dict[int, str]) -> str:
+        """Return a PostScript encoding array for the encoding."""
+        result = [
+            '/Encoding 256 array\n0 1 255 { 1 index exch /.notdef put} for'
+        ]
+        result.extend(
+            f'dup {i} /{glyph} put'
+            for i, glyph in sorted(encoding.items())
+            if glyph != '.notdef'
+        )
+        result.append('readonly def\n')
+        return '\n'.join(result)
```
Probably can join these:
```suggestion
        return '\n'.join([
            '/Encoding 256 array\n0 1 255 { 1 index exch /.notdef put} for',
            *(
                f'dup {i} /{glyph} put'
                for i, glyph in sorted(encoding.items())
                if glyph != '.notdef'
            ),
            'readonly def\n',
        ])
```
Maybe the same for the rest.

jkseppan replied:
Thanks!

QuLogic commented on lib/matplotlib/backends/backend_pdf.py:
```diff
@@ -1045,7 +1037,15 @@ def _embedTeXFont(self, fontinfo):
+        widths = [(1000 * metrics.tex_width) >> 20
+                  if (metrics := tfm.get_metrics(char)) else 0
+                  for char in range(fc, lc + 1)]
+        fontdict['Widths'] = (widthsObject := self.reserveObject('glyph widths'))
```
This walrus operator appears superfluous.
```suggestion
        fontdict['Widths'] = widthsObject = self.reserveObject('glyph widths')
```

QuLogic commented on lib/matplotlib/tests/test_usetex.py:
```diff
@@ -156,6 +157,71 @@ def test_missing_psfont(fmt, monkeypatch):
 
 
+def test_pdf_type1_font_subsetting():
+    """Test that fonts in PDF output are properly subsetted"""
```
```suggestion
    """Test that fonts in PDF output are properly subset."""
```

jkseppan commented on doc/users/next_whats_new/type1_subset.rst:
```diff
@@ -0,0 +1,6 @@
+Type 1 fonts are now subsetted in PDF output
+--------------------------------------------
```
I think I disagree here... the English verb "set" is irregular in that way, but if you search for "subsetted" in the context of fonts, it seems to be fairly common, including in fonttools and various Adobe forums. See also the accepted answer to [this question](https://english.stackexchange.com/questions/297323/simple-past-of-the-verb-subset) and possibly the discussion of *flied out* in Steven Pinker's *Words and Rules*.

jklymak replied:
Maybe just be a bit more verbose?  

```suggestion
PDFs embed just the subset of Type 1 glyphs that are used
-----------------------------------------------------------
```

QuLogic replied:
I don't disagree that this is the correct conjugation in the past tense, rather that these sentences are not in the past tense. It is stating what is and in the (foreseeable) future shall occur.

jkseppan replied:
I reworded the whole paragraph to be hopefully more understandable on its own.

jkseppan commented:
I was trying to improve test coverage and discovered some cases where this implementation breaks the output. Math examples seem to work fine, but there are some text fonts that use less common features that I'm clearly not handling right. Here are two cases:

```py
@image_comparison(["subsetting-heuristica.pdf"])
def test_subsetting_heuristica():
    # Heuristica uses the callothersubr operator for some glyphs
    mpl.rcParams['text.latex.preamble'] = '\n'.join((
        r'\usepackage{heuristica}',
        r'\usepackage[T1]{fontenc}',
        r'\usepackage[utf8]{inputenc}'
    ))
    fig, ax = plt.subplots()
    ax.text(0.1, 0.1, r"BHTem", usetex=True, fontsize=50)
    ax.text(0.1, 0.3, "fi", usetex=True, fontsize=50)
    ax.text(0.1, 0.5, "ffl", usetex=True, fontsize=50)
    ax.set_xticks([])
    ax.set_yticks([])


@image_comparison(["subsetting-dejavusans.pdf"])
def test_subsetting_dejavusans():
    # DejaVuSans uses the seac operator to compose characters with diacritics
    mpl.rcParams['text.latex.preamble'] = '\n'.join((
        r'\usepackage{DejaVuSans}',
        r'\usepackage[T1]{fontenc}',
        r'\usepackage[utf8]{inputenc}'
    ))

    fig, ax = plt.subplots()
    ax.text(0.1, 0.1, r"\textsf{ñäö}", usetex=True, fontsize=50)
    ax.text(0.1, 0.3, "fi", usetex=True, fontsize=50)
    ax.text(0.1, 0.5, "ffl", usetex=True, fontsize=50)
    ax.set_xticks([])
    ax.set_yticks([])
```

The Heuristica callothersubr feature actually doesn't seem broken, but the fi and ffl ligatures are lost in both fonts.

<img width="429" alt="image" src="https://github.com/user-attachments/assets/a344ae59-a713-4458-9bac-81f7f5470e40" />

<img width="353" alt="image" src="https://github.com/user-attachments/assets/4c28e154-9a8b-4b71-80e8-2ee3095857b9" />

Both work without subsetting, although the metrics for the ligature glyphs seem to be wrong in the current code.

I'll see if I can figure out what's wrong.

jkseppan commented:
The ligature problem is probably because we don't apply the encoding from TeX's font configuration to the font before subsetting. The custom encoding array is output in the PDF file but should also be used to map from character codes to glyph names. The seac issue might be a different encoding problem where we should do the lookups using Adobe Standard Encoding and not the font's own encoding.

jkseppan commented:
It seems that some of the latest changes broke compatibility with older GhostScript again.

But while I debug that, a note about the new tests: they use font packages that are available on Debian or Ubuntu only by installing texlive-fonts-extra, which brings in a lot of other fonts too. Currently these tests get skipped on all runners, but would it make sense to install the extra fonts on just one of the runners to allow these tests to get run somewhere?

jkseppan commented:
I added a test using Bitstream Charter, which is part of texlive-fonts-recommended, so we get at least some coverage of the full Type-1 subsetting code path.

I fixed the gs compatibility issue, which was about a broken Encoding object.

anntzer commented on lib/matplotlib/_type1font.py:
```diff
@@ -768,13 +832,279 @@ def transform(self, effects):
+        lenIV = self.prop.get('lenIV', 4)
+        encrypted = [
+            self._encrypt(charstrings[glyph], 'charstring', lenIV).decode('latin-1')
+            for glyph in glyphs
```
Should this (and _subset_subrs below) sort the glyphs (and subrs) to ensure reproducibility?  (as set ordering changes over runs)

jkseppan replied:
The subrs are already in order (the loop is `for i in range(n_subrs)`) but sorting the glyphs is a good idea.

jkseppan commented:
I removed the extra type annotations, which were incomplete in any case. I'll make a separate PR to annotate the entire file.

jklymak commented on doc/users/next_whats_new/type1_subset.rst:
```diff
@@ -0,0 +1,9 @@
+The fonts that get used are usually "Type 1" fonts.
+They used to be embedded in full
+but are now limited to the glyphs that are actually used in the figure.
+This reduces the size of the resulting PDF files.
```
This reads well to me.  Thanks!

tacaswell commented:
Will these new test be adjusted by #29816 ?  If so we should sequence that one first.

jkseppan commented:
> Will these new test be adjusted by https://github.com/matplotlib/matplotlib/pull/29816 ? If so we should sequence that one first.

I don't think these depend on FreeType, since the usetex case uses TeX for layout and parses dvi files to determine the coordinates of glyphs.

jkseppan commented:
The earlier CI error seems to have been caused by a newer mypy version detecting some more types in an unrelated file. The fix is also in PR #30119.

QuLogic commented on lib/matplotlib/backends/backend_pdf.py:
```diff
@@ -1038,9 +1039,32 @@ def _embedTeXFont(self, fontinfo):
+                  for char in range(fc, lc + 1)]
+        fontdict['Widths'] = widthsObject = self.reserveObject('glyph widths')
         self.writeObject(fontdictObject, fontdict)
+        self.writeObject(widthsObject, widths)
```
Does this need to move after the `fontdict`? It should be reserved already.

jkseppan replied:
It doesn't matter which order we write them in, but I think the Widths object needs to be indirect and not included in the font dictionary. This does need to come after we know the font encoding, which in turn depends on the glyph set.

But I guess you mean that this would read better if we write out the widths object first? Either way works for me.

QuLogic replied:
Yes, it did used to be written first as well, but moreso that it would be along with the other width-related code.

QuLogic commented on lib/matplotlib/backends/backend_pdf.py:
```diff
@@ -767,8 +765,6 @@ def __init__(self, filename, metadata=None):
 
     fontNames = _api.deprecated("3.11")(property(lambda self: self._fontNames))
     dviFontInfo = _api.deprecated("3.11")(property(lambda self: self._dviFontInfo))
-    type1Descriptors = _api.deprecated("3.11")(
-        property(lambda self: self._type1Descriptors))
```
This isn't private, so we need to keep the property for the deprecation, even if it just returns an empty dictionary.

jkseppan commented:
> It looks like you'll need to install `heuristica` and `DejaVuSans` texlive packages in `.github/workflows/tests.yml` to get those new tests running.

Yes, but Debian (and hence Ubuntu) bundles them into `texlive-fonts-extra` with a large number of other fonts. Not sure that slowing down all test runners is worth it, but perhaps we could do it in just one runner?

QuLogic commented on .github/workflows/tests.yml:
```diff
@@ -175,6 +176,15 @@ jobs:
           esac
 
+      - name: Install TeX Live dependencies
+        run: |
+          case "${{ runner.os }}" in
+          Linux)
+            mkdir -p $HOME/texmf
+            tlmgr init-usertree
+            tlmgr install --force dejavu heuristica
+          esac
```
```suggestion
        if: startsWith(runner.os, 'Linux')
        run: |
          mkdir -p $HOME/texmf
          tlmgr init-usertree
          tlmgr install --force dejavu heuristica
```

jkseppan replied:
Yeah, it bails out with

    tlmgr: Local TeX Live (2021) is older than remote repository (2025).

QuLogic commented:
Here are the times for the "Install OS dependencies" step in the previous 3 commits:
| Commit                                   | 3.11 minver | 3.11  | 3.12  | 3.13  | 3.13 freethread | 3.12 arm |
| ---------------------------------------- | ----------- | ----- | ----- | ----- | --------------- | -------- |
| 5006acc7f0ae1e29019e5afaf114d62e3ba3e999 | 1m9s        | 1m21s | 2m59s | 1m52s | 1m24s           | 3m54s    |
| 546a7671287e049780997d73840ea85bc0c288ca | 1m10s       | 1m24s | 2m23s | 1m52s | 1m22s           | 3m19s    |
| 5c2f6bdd5f8e2807986bc33564992e04b0471e14 | 1m13s       | 1m15s | 2m16s | 2m42  | 1m15s           | 2m20s    |

And for the 2 last commits with the additional 1 or 2 packages:
| Commit                                   | 3.11 minver | 3.11  | 3.12  | 3.13  | 3.13 freethread | 3.12 arm |
| ---------------------------------------- | ----------- | ----- | ----- | ----- | --------------- | -------- |
| 7741e9b5ca4d652679acb47ad4719c263292342e | 2m31s       | 2m36s | 5m3s  | 2m50s | 2m10s           | 3m25s    |
| f411fb383c5a50c54167d22fea03abe5502a52ad | 2m30s       | 2m58s | 2m56s | 2m34s | 2m5s            | 5m15s    |

It looks like it would add about ~1m per job?

jkseppan commented:
Some ideas:

- install TeX without using the Ubuntu packages, with e.g. TinyTeX (not guaranteed to be faster)
- run the tests inside Docker, build images in phases so the base OS and TeX are installed in advance, store the images in GitHub Container Registry (can almost certainly shave off several minutes off the build time, at the cost of some complexity and may need to pay GitHub for the GCR storage)
- set up TeX only in a subset of the test runners, or dedicate some runners to TeX tests only (the latter would increase parallelism and might reduce wall-clock time, but Amdahl's law applies)

tacaswell commented:
I think option 3 (only run on a subset).  I would go with the newest Python.

QuLogic commented:
Should be able to set an optional addition in the `matrix` just like we do with `matrix.extra-requirements`.

jkseppan commented:
I moved the packages texlive-latex-extra, texlive-fonts-extra and texlive-lang-cyrillic to one runner only. (The texlive-latex-extra package has been included for a long time and I'm not sure which exact LaTeX dependencies are used by our tests.)

Complete run times:

Job | Run time |  
-- | -- | --
Python 3.11 on ubuntu-22.04 (Minimum Versions) | 18m 37s |  
Python 3.11 on ubuntu-22.04 | 17m 31s |  
Python 3.12 on ubuntu-22.04-arm | 12m 22s |  
Python 3.13 on ubuntu-22.04 (Extra TeX packages) | 20m 25s |  
Python 3.13t on ubuntu-22.04 Free-threaded | 16m 58s |  
Python 3.12 on ubuntu-24.04 | 20m 3s |  
Python 3.11 on macos-13 | 24m 18s |  
Python 3.12 on macos-14 | 17m 44s |  
Python 3.13 on macos-14 | 16m 36s |  
Create issue on failure | 0s |  
  | 2h 44m 34s |  


For comparison, a [recent commit on main](https://github.com/matplotlib/matplotlib/actions/runs/15481096007/usage):


Job | Run time |  
-- | -- | --
Python 3.11 on ubuntu-22.04 (Minimum Versions) | 19m 24s |  
Python 3.11 on ubuntu-22.04 | 17m 10s |  
Python 3.12 on ubuntu-22.04-arm | 11m 15s |  
Python 3.13 on ubuntu-22.04 | 18m 24s |  
Python 3.13t on ubuntu-22.04 Free-threaded | 16m 54s |  
Python 3.12 on ubuntu-24.04 | 20m 32s |  
Python 3.11 on macos-13 | 22m 39s |  
Python 3.12 on macos-14 | 16m 5s |  
Python 3.13 on macos-14 | 15m 9s |  
Create issue on failure | 0s |  
  | 2h 37m 32s |  


In this case it seems to have added a little under two minutes to that runner, including both the time taken to install and the extra tests run. There seems to be quite a bit of variation from run to run, and most of the overall increase is due to the macos builders where I don't think we even install TeX.

QuLogic commented:
https://matplotlib.org/devdocs/install/dependencies.html#latex-dependencies says it's for `import`, `sfmath`, and `type1cm`, and `type1cm` is used unconditionally in the preamble. According to the log, `texlive-latex-extra` is already being installed by dependency of something else which is why it didn't break, so I don't think it makes sense to move it to the extras job only. It should stay for explicitness in all jobs.

QuLogic commented:
> Closes #127.

Congrats on closing a 3-digit issue.
